dbms/src/Interpreters/Join.cpp

```diff
@@ -288,6 +348,33 @@ void Join::setSampleBlock(const Block & block)
             convertColumnToNullable(sample_block_with_columns_to_add.getByPosition(i));
 }
 
+void Join::TSRowRef::insert(Join::ASOFTimeType t, const Block * block, size_t row_num)
+{
+    ts.insert(std::pair(t, RowRef(block, row_num)));
+}
+
+std::string Join::TSRowRef::dumpStructure() const
+{
+    std::stringstream ss;
+
+    for (auto const& x : ts)
+    {
+        ss << "(t=" << x.first << " row_num=" << x.second.row_num << " ptr=" << x.second.block << "),";
+    }
+
+    return ss.str();
+}
+size_t Join::TSRowRef::size() const
+{
+    return ts.size();
+}
+std::optional<std::pair<Join::ASOFTimeType, Join::RowRef>> Join::TSRowRef::find_asof(Join::ASOFTimeType t) const
+{
+    auto it = ts.upper_bound(t);
+    if (it == ts.cbegin())
+        return {};
+    return *(--it);
```

Comment:
upper_bound cannot return an iterator before begin()
As an example program demonstrating the functionality, which works with both the insertion and it commented out.
```
#include <iostream>
#include <map>

int main()
{
    auto x = std::map<int,int>();
    x[3] = 5;
    
    for(int i = 0; i < 10; i++) {
        std::cout << i << " ";
        auto it = x.upper_bound(i);
        if (it == x.cbegin()) {
            std::cout << "not found" << std::endl;
            continue;
        }
        auto val = *(--it);
        std::cout << "val=" << val.second << std::endl;
    }
}
```

I agree that the ON join condition isn't properly supported yet with ASOF - I haven't had a look through the code path that that takes, will report back on that. However, it seems that at the point that you get to the Join class you only know about the column names that you want to join on and not how you want to join them (eg. bigger or smaller) so it would probably require quite a lot of code plumbing to get those relationships usable in the Join object.

Reply:
Cool if so. Lets's add asof join with 0 columns an 1 colunt into our tests. Just to be calm.

Reply:
Done 9fc7f67c5